builtin/logical/ad/config/engineconf.go

```diff
@@ -1,56 +1,43 @@
 package config
 
 import (
-	"errors"
-
 	"github.com/hashicorp/vault/helper/activedirectory"
 )
 
-// newUnsetEngineConf provides an engineConf that's essentially empty,
-// but won't cause a panic.
-// If `engineConf.IsSet == false`, that reflects that the user doesn't
-// currently have any configuration stored.
-func newUnsetEngineConf() *EngineConf {
-	return &EngineConf{
-		PasswordConf: &PasswordConf{},
-		ADConf:       &activedirectory.Configuration{},
-		IsSet:        false,
-	}
-}
-
-func newEngineConf(passwordConf *PasswordConf, adConf *activedirectory.Configuration) (*EngineConf, error) {
-
-	// usernames and passwords aren't required for the AD client in general,
-	// but for this backend, we need them for password rotation.
-	if adConf.Username == "" || adConf.Password == "" {
-		return nil, errors.New("a username and password must be provided to perform password rotation")
-	}
-
-	return &EngineConf{
-		PasswordConf: passwordConf,
-		ADConf:       adConf,
-		IsSet:        true,
-	}, nil
-}
-
-type EngineConf struct {
+type engineConf struct {
 	PasswordConf *PasswordConf
 	ADConf       *activedirectory.Configuration
-
-	// IsSet reflects whether the *user* has set the configuration.
-	IsSet bool
 }
 
-func (c *EngineConf) Map() map[string]interface{} {
+// Since *engineConf will be nil if it's unset by the user
+// or if its cached version has been invalidated,
+// let's be super defensive around nil pointers here.
+func (c *engineConf) Map() map[string]interface{} {
 	combined := make(map[string]interface{})
-	if !c.IsSet {
+	if c == nil {
 		return combined
 	}
-	for k, v := range c.PasswordConf.Map() {
-		combined[k] = v
+	if c.PasswordConf != nil {
+		for k, v := range c.PasswordConf.Map() {
+			combined[k] = v
+		}
 	}
-	for k, v := range c.ADConf.Map() {
-		combined[k] = v
+	if c.ADConf != nil {
+		for k, v := range c.ADConf.Map() {
+			combined[k] = v
+		}
 	}
 	return combined
 }
+
+func (c *engineConf) Immutable() ImmutableEngineConf {
```

Comment:
Can you explain the need for a new type and Immutable() method?

Reply:
Yeah! Because if I return a non-pointer to an EngineConf in an effort to make it immutable, it still contains the pointers to the ADConf and the PasswordConf, so it's still totally mutable. So I made an immutable version of it that also held non-pointers for those sub-objects.

Reply:
I don't think you need to get an immutable copy of the object. In the case of read, what we need is a guaranteed point-in-time representation of the object, so a mutex whenever the object is accessed should suffice. The lock guarantees that the object cannot be changed during the time it's being locked (i.e. during the read call).

It looks like you're already locking with `m.rwMutex` in `m.Config()` so I think you can safely remove this method and `ImmutableEngineConf`.

Reply:
Thanks Jim and Calvin!

Upon your feedback and reflection, I'll strip the Immutable stuff. It should be safe to simply return `*EngineConf` outside the package. As long as it's only read, there won't be any races. If anyone writes to it, it'll be caught by the race detector before it's added to any PR.